### ISSUE TITLE
Overflow error in Commit Dialog

### ISSUE THREAD

**Issue by palver123** [gitextensions/gitextensions#7023]
## Current behaviour

WinForms control in commit dialog crashes.

## Expected behaviour

Do not crash.

## Steps to reproduce

Start changing the encodings (Unicode might be what causes the error). I was viewing a diff of an unstaged file (a C# source file, that I must not attach). I was changing the encoding because the diff showed something fishy at the start of the file (near the using statements).

I could reproduce the issue as many times as I wanted by selecting 'Unicode' encoding.

## Error Details
```
System.OverflowException: Overflow error.
   at System.Drawing.Graphics.CheckErrorStatus(Int32 status)
   at System.Drawing.Graphics.FillRectangle(Brush brush, Single x, Single y, Single width, Single height)
   at System.Drawing.Graphics.FillRectangle(Brush brush, RectangleF rect)
   at ICSharpCode.TextEditor.TextView.DrawDocumentWord(Graphics g, String word, Point position, Font font, Color foreColor, Brush backBrush)
   at ICSharpCode.TextEditor.TextView.PaintLinePart(Graphics g, Int32 lineNumber, Int32 startColumn, Int32 endColumn, Rectangle lineRectangle, Int32 physicalXPos)
   at ICSharpCode.TextEditor.TextView.PaintDocumentLine(Graphics g, Int32 lineNumber, Rectangle lineRectangle)
   at ICSharpCode.TextEditor.TextView.Paint(Graphics g, Rectangle rect)
   at ICSharpCode.TextEditor.TextArea.OnPaint(PaintEventArgs e)
   at System.Windows.Forms.Control.PaintWithErrorHandling(PaintEventArgs e, Int16 layer)
   at System.Windows.Forms.Control.WmPaint(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Environment
- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.22.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3416.0
- DPI 96dpi (no scaling)

**Comment by palver123** [comment]
It might be an error in ICSharpCode TextEditor I don't know.

**Comment by gerhardol** [comment]
Duplicate of #6633
That issue has some more reproduce info too.

**Comment by palver123** [comment]
I am leaving for a holiday, but after that I might debug into it to find out more.

**Comment by palver123** [comment]
My suscpicion:
`DiffViewerLineNumberControl.Width` returns -2147483599 if `MaxLineNumber `is 0 (because of log 0). This number is out of the [supported bounds of GDI](https://stackoverflow.com/questions/3468495/what-are-the-hard-bounds-for-drawing-coordinates-in-gdi) (and thus Windows Forms controls). This initiates the following avalanche:

 - TextArea.OnPaint() is called.
 - It calculates `currentXPos `from `leftMargins`
``` csharp
foreach (var margin in leftMargins)
    if (margin.IsVisible)
    {
        // margin.Width returns nonsense!!!
        var marginRectangle = new Rectangle(currentXPos, currentYPos, margin.Width, Height - currentYPos);
        if (marginRectangle != margin.DrawingPosition)
        {
            ...
            margin.DrawingPosition = marginRectangle; // Bad value!
        }

        currentXPos += margin.DrawingPosition.Width;
        ...
    }
```
 - DiffViewerLineNumberControl.DrawingPosition is set to a nonsense value. So is currentXPos . This later causes a GDI OverflowException here:
``` csharp
// currentXPos is nonsense!
var textViewArea = new Rectangle(currentXPos, currentYPos, Width - currentXPos, Height - currentYPos);
if (textViewArea != TextView.DrawingPosition)
{
    ...
    TextView.DrawingPosition = textViewArea; // Bad value!
    ...
}

if (clipRectangle.IntersectsWith(textViewArea))
{
    textViewArea.Intersect(clipRectangle);
    if (!textViewArea.IsEmpty)
        TextView.Paint(g, textViewArea); // textViewArea is nonsense
}
```

**Comment by palver123** [comment]
DiffLineNumAnalyzer is the class that calculate MaxLineNumber.  In my case the following code gets called only once:

``` csharp
public void Add(DiffLineInfo diffLine)
{
    _diffLines.Add(diffLine.LineNumInDiff, diffLine);
    MaxLineNumber = Math.Max(MaxLineNumber,
        Math.Max(diffLine.LeftLineNumber, diffLine.RightLineNumber));
}
```

and both `LeftLineNumber `and `RightLineNumber `are set to `DiffLineInfo.NotApplicableLineNum`,

**Comment by palver123** [comment]
Probably should change the current getter:

``` csharp
if (_visible && _diffLines.Any())
{
    int maxDigits = (int)Math.Log10(MaxLineNumber) + 1;
    return ...;
}
```

to

``` csharp
if (_visible && _diffLines.Any())
{
    int maxDigits = MaxLineNumber == 0 ? 0 : ((int)Math.Log10(MaxLineNumber) + 1);
    return ...;
}
```

**Comment by RussKie** [comment]
Are you able to send a PR with a fix?

**Comment by palver123** [comment]
Not until next Friday, cuz I am not near a computer until then. If you want to include this in a release until Friday, you can commit it yourself.

**Comment by palver123** [comment]
I was wondering about this issue and I have the following worries:

I have found a bug regarding `DiffViewerLineNumberControl.Width` in case the displayed file _has 0 lines_. For example an empty file, right? However, the bug does not occur in case of an empty file (probably would have been fixed long time ago then). I could only reproduce it with a file that has ~250 lines (and using Unicode encoding!)

This make me wonder if there is another error somewhere waiting to be fixed. I did find a solution for my particular case but it feels like a band-aid and not a real fix. I'd like to go deeper to find out how Unicode encoding results in GitExtensions believing that my ~250 lines long file _has 0 lines_.

As soon as I find out more we can tell if it's worth opening a new issue or continue using this issue. Or I find no errors and can just close this issue and create a pull request containing the fix for `DiffViewerLineNumberControl.Width`.

**Comment by RussKie** [comment]
Thank you

**Comment by palver123** [comment]
> Why does not DiffViewerLineNumberControl.Width cause a crash when viewing an empty file?

Because for an empty file `DiffViewerLineNumberControl._diffLines` is empty and get_Width returns 0.

> Why does GitExtensions think that my ~250 lines long file is empty when I used Unicode encoding in the diff viewer?

Because `DiffLinesInfo.MaxLineNumber` is initialized to 0. My diff patch looks like this with Unicode encoding:

![image](https://user-images.githubusercontent.com/12460850/63724226-93ddb000-c857-11e9-83f5-42c61f032c33.png)

As you can see there are no diff hunks in the patch file so `DiffLinesInfo.MaxLineNumber` will never increase. I think there is no problem with this behaviour.

I am making a pull request and close this issue.
